File: android/localasset-core/src/main/kotlin/io/github/xesam/android/localasset/core/loader/CompositeResourceLoader.kt
package io.github.xesam.android.localasset.core.loader

import io.github.xesam.android.localasset.core.api.ResourceLoader
import io.github.xesam.android.localasset.core.error.ResourceErrorCategory
import io.github.xesam.android.localasset.core.error.ResourceException
import io.github.xesam.android.localasset.core.api.TypedResourceLoader
import io.github.xesam.android.localasset.core.model.ResourceData
import io.github.xesam.android.localasset.core.model.ResourceDescriptor

class CompositeResourceLoader(
    private val loaders: List<ResourceLoader>,
) : ResourceLoader {
    override fun load(descriptor: ResourceDescriptor): ResourceData {
        val typedLoaders = loaders.mapNotNull { it as? TypedResourceLoader }
        val loader = typedLoaders.firstOrNull { it.canLoad(descriptor.source) }
            ?: run {
                val nonTyped = loaders.filter { it !is TypedResourceLoader }
                val nonTypedNames = nonTyped.map { it::class.simpleName ?: it::class.toString() }
                throw ResourceException(
                    ResourceErrorCategory.LOAD_ERROR,
                    "no loader for source type ${descriptor.source::class.simpleName}" +
                        if (nonTyped.isNotEmpty()) {
                            " (${typedLoaders.size} typed loader(s) checked; " +
                                "${nonTyped.size} loader(s) do not implement TypedResourceLoader: $nonTypedNames)"
                        } else {
                            " (${typedLoaders.size} typed loader(s) checked)"
                        },
                )
            }
        return loader.loadTyped(descriptor)
    }
}

File: ios/localasset-core/Sources/LocalAssetCore/loader/CompositeResourceLoader.swift
public class CompositeResourceLoader: ResourceLoader {
    private let loaders: [ResourceLoader]

    public init(loaders: [ResourceLoader]) {
        self.loaders = loaders
    }

    public func load(descriptor: ResourceDescriptor) throws -> ResourceData {
        let typed = loaders.compactMap { $0 as? TypedResourceLoader }
        if let loader = typed.first(where: { $0.canLoad(source: descriptor.source) }) {
            return try loader.load(descriptor: descriptor)
        }
        let nonTypedNames = loaders
            .filter { !($0 is TypedResourceLoader) }
            .map { String(describing: type(of: $0)) }
        throw ResourceException(
            .loadError,
            "no loader for source type \(descriptor.source)" +
                (nonTypedNames.isEmpty
                    ? " (\(typed.count) typed loader(s) checked)"
                    : " (\(typed.count) typed loader(s) checked; \(nonTypedNames.count) loader(s) do not implement TypedResourceLoader: \(nonTypedNames))")
        )
    }
}
